### PR TITLE
Enabling GTPv1-U

### PR DESCRIPTION
**Pull request by ifilippov**
No Testing, even dumping. Just prototype.

### CODE REVIEW

File: packet/gtp.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+// functions are base variants without any checkings which can be
+// painlessly changed for exact situation.
+
+package packet
+
+import (
+	"fmt"
+	"unsafe"
+
+	. "github.com/intel-go/yanff/common"
```
Comment (aregm) [inline packet/gtp.go]: Better to do alias and we need to solve the "common.go" problem

File: packet/gtp.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+)
+
+// GTPv1-U and GTPv1-C
+type GTPHdr struct {
+	HeaderType          uint8 // version, protocol type, extension header/sequence number/N-PDU flags
+	MessageType         uint8
+	MessageLength       uint16
+	TEID                uint32
+	SequenceNumber      uint16 // this is valid only with sequence number flag
```
Comment (aregm) [inline packet/gtp.go]: remove "this is"

File: packet/gtp.go (hunk 1 above)
Comment (Quasilyte) [inline packet/gtp.go]: Usage of dot imports outside of testing code is discouraged:
See https://github.com/golang/go/wiki/CodeReviewComments#import-dot.

Comment (Quasilyte) [reply packet/gtp.go]: Many go static code analyzers can't handle dot-imports.
Also, `go lint` will say: `should not use dot imports`.

**Comment by ifilippov** [comment]
Quasilyte: We use dot import from common.go across whole framework due to significant code reducing. Maybe it can be refactored somehow but not in this commit.

**Comment by Quasilyte** [comment]
You mentioned that this is "prototyping", so it was quite wrong to comment on this.
I done this before reading the actual description, sorry.

File: packet/gtp.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	EchoResponse                          = 2
+	SupportedExtensionHeadersNotification = 31
+	// Signalling messages: tunnel management
+	ErrorIndication = 26
+	EndMarker       = 254
+	// G-PDU message = encapsulated user data
+	G_PDU = 255
+)
+
+// TODO Add "Information Element@ constants
```
Comment (aregm) [inline packet/gtp.go]: What does comment want to say?

File: packet/gtp.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	Length              uint8 // in 4 octets, here always 0x01
+	PDCP_PDU_Number_1   uint8
+	PDCP_PDU_Number_2   uint8
+	NextExtensionHeader uint8
+}
+
+// TODO for GTP-U only
+// MessageType values
+const (
+	// Signalling messages: path management
```
Comment (aregm) [inline packet/gtp.go]: Alignment

File: packet/gtp.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+		min += fmt.Sprintf(`, Sequence number %d`, SwapBytesUint16(hdr.SequenceNumber))
+	}
+	if hdr.HeaderType&0x01 != 0 {
+		min += fmt.Sprintf(`, N-PDU number %d`, hdr.NPDUNumber)
+	}
+	if hdr.HeaderType&0x04 != 0 {
+		min += fmt.Sprintf(`, N-PDU number %d`, hdr.NextExtensionHeader)
+		// TODO add dumping of all extension headers
+	}
+	return min
```
Comment (aregm) [inline packet/gtp.go]: Why name 'min'?

File: packet/gtp.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// EncapsulateIPv4GTP assumes that user wants to build ether->IPv4->UDP->GTP->payload data structure
+// with standart IPv4 header size. It is also assumed that payload type is IPv4, so no etherType changes are needed
+func (packet *Packet) EncapsulateIPv4GTP(TEID uint32) bool {
+	if !packet.EncapsulateHead(EtherLen, IPv4MinLen+UDPLen+GTPMinLen) {
+		return false
+	}
+	gtp := (*GTPHdr)(unsafe.Pointer(uintptr(packet.unparsed()) + IPv4MinLen + UDPLen))
+	gtp.HeaderType = 0x30   // 001 - GTPv1, 1 - not GTP', 0 - reserved, 000 - no optional fields
```
Comment (aregm) [inline packet/gtp.go]: Better to define constant

File: packet/gtp.go (hunk 4 above)
Comment (ifilippov) [reply packet/gtp.go]: It is comment for underline variable

File: packet/gtp.go (hunk 3 above)
Comment (ifilippov) [reply packet/gtp.go]: That we need to add "Information Element" constants which are defined in protocol RFCs

File: packet/gtp.go (hunk 5 above)
Comment (ifilippov) [reply packet/gtp.go]: Because it is minimal protocol length

**Pull request merged**